gson/src/main/java/com/google/gson/internal/bind/TreeTypeAdapter.java

```diff
@@ -83,6 +83,7 @@ public TreeTypeAdapter(JsonSerializer<T> serializer, JsonDeserializer<T> deseria
   }
 
   private TypeAdapter<T> delegate() {
+    // Allows racy initialization of `delegate` by multiple threads
```

Comment:
More than one thread can observe `d == null` and therefore call `gson.getDelegateAdapter(...)`. That's _probably_ not a problem? Presumably multiple calls will return the same or at least an equivalent result, and one of them will end up remaining in `delegate`.

Reply:
Yes exactly, that is what I meant with the comment, to clarify that this case has been considered.
Maybe for some rare corner cases it might make a difference, but I assume for the majority of the use cases it does not make a difference.
However, if you want I can also replace this with double checked locking to be on the safe side.

Reply:
I think it's fine as it stands but the comment could be a bit clearer. For example:

```java
// A race might lead to `delegate` being assigned by multiple threads but the last assignment will stick.
```